File: Count of Substrings Containing Every Vowel and K Consonants II/Count of Substrings Containing Every Vowel and K Consonants II.java
// Solution for Count of Substrings Containing Every Vowel and K Consonants II in JAVA

class Solution {
    public long countOfSubstrings(String word, int k) {
        int n = word.length();
        Set<Character> vowels = new HashSet<>(Arrays.asList('a', 'e', 'i', 'o', 'u'));
        Map<Character, Integer> mp = new HashMap<>(); // vowelsCount
        int[] nextConsonant = new int[n];
        int lastConsIdx = n; // consonant index
        for (int i = n - 1; i >= 0; i--) {
            nextConsonant[i] = lastConsIdx;
            if (!vowels.contains(word.charAt(i))) {
                lastConsIdx = i;
            }
        }
        int i = 0, j = 0, cons = 0;
        long res = 0;
        while (j < n) {
            if (vowels.contains(word.charAt(j))) {
                mp.put(word.charAt(j), mp.getOrDefault(word.charAt(j), 0) + 1);
            } else {
                cons++;
            }
            while (cons > k) {
                char ch = word.charAt(i);
                if (vowels.contains(ch)) {
                    mp.put(ch, mp.get(ch) - 1);
                    if (mp.get(ch) == 0) mp.remove(ch);
                } else {
                    cons--;
                }
                i++;
            }
            while (i < n && mp.size() == 5 && cons == k) {
                res += (nextConsonant[j] - j);
                char ch = word.charAt(i);
                if (vowels.contains(ch)) {
                    mp.put(ch, mp.get(ch) - 1);
                    if (mp.get(ch) == 0) mp.remove(ch);
                } else {
                    cons--;
                }
                i++;
            }
            j++;
        }
        return res;
    }
}

File: Count of Substrings Containing Every Vowel and K Consonants II/Count of Substrings Containing Every Vowel and K Consonants II.py
# Solution for Count of Substrings Containing Every Vowel and K Consonants II in PY

class Solution:
    def countOfSubstrings(self, word: str, k: int) -> int:
        vowels = {'a', 'e', 'i', 'o', 'u'}
        mp = {} # vowelsCount
        n = len(word)
        nextConsonant = [n] * n
        lastConsIdx = n # consonant index
        for i in range(n - 1, -1, -1):
            nextConsonant[i] = lastConsIdx
            if word[i] not in vowels:
                lastConsIdx = i
        i = j = cons = res = 0
        while j < n:
            if word[j] in vowels:
                mp[word[j]] = mp.get(word[j], 0) + 1
            else:
                cons += 1
            while cons > k:
                ch = word[i]
                if ch in vowels:
                    mp[ch] -= 1
                    if mp[ch] == 0:
                        del mp[ch]
                else:
                    cons -= 1
                i += 1
            while i < n and len(mp) == 5 and cons == k:
                res += (nextConsonant[j] - j)
                ch = word[i]
                if ch in vowels:
                    mp[ch] -= 1
                    if mp[ch] == 0:
                        del mp[ch]
                else:
                    cons -= 1
                i += 1
            j += 1
        return res
